### PR TITLE
SSDP with XML to make SABnzbd visible on Windows

### PR DESCRIPTION
**Pull request by sanderjo**
As discussed https://github.com/sabnzbd/sabnzbd/issues/1641

### CODE REVIEW

File: sabnzbd/utils/ssdp.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    def stop(self):
+        logging.info("Stopping SSDP")
+        self.__stop = True
+
+    def run(self):
+        logging.info("Serving SSDP on %s as %s", self.__host, self.__server_name)
+        logging.info("self.__url is %s", self.__url)
+
+        # warning ... hack ahead ... to be solved
```
Comment (sanderjo) [inline sabnzbd/utils/ssdp.py:46]: @Safihre I need your help here ...

File: sabnzbd/utils/ssdp.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                sock.sendto(mySSDPbroadcast, (MCAST_GRP, MCAST_PORT))
+            time.sleep(2)
+
+    def serve_xml(self):
+        """Returns an XML-structure based on the information being
+        served by this service, returns nothing if not running"""
+        if self.__stop:
+            return
+        # Use self.__host, self.__url, self.__server_name to do stuff!
+        logging.debug("description.xml was retrieved by ...")
```
[the remark this answers is not in the document]
Comment (Safihre) [reply sabnzbd/utils/ssdp.py:99]: Yes, in `interface.py` in the new function you can add logging by accessing the `cherrypy.request` object. Try it in Debug-mode. 
For example there is `cherrypy.request.remote.ip` and more.

File: sabnzbd/utils/ssdp.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+<minor>0</minor>
+</specVersion>
+<URLBase>{self.__url}</URLBase>
+<device>
+<deviceType>urn:schemas-upnp-org:device:Basic:1</deviceType>
+<friendlyName>SABnzbd ({self.__host})</friendlyName>
+<manufacturer>SABnzbd Team</manufacturer>
+<manufacturerURL>http://www.sabnzbd.org</manufacturerURL>
+<modelDescription>SABnzbd downloader</modelDescription>
+<modelName>SABnzbd 3.4.5</modelName>
```
[the remark this answers is not in the document]
Comment (Safihre) [reply sabnzbd/utils/ssdp.py:116]: Add new variables to the `SSDP` object and the `start_ssdp` function. I just added the current `self.__xx` ones, but feel free to add more!

File: sabnzbd/utils/ssdp.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        logging.info("self.__url is %s", self.__url)
+
+        descriptionxmlURL = self.__url + "/description.xml"
+
+        # the standard multicast settings for SSDP:
+        MCAST_GRP = "239.255.255.250"
+        MCAST_PORT = 1900
+        MULTICAST_TTL = 2
+
+        mySSDPbroadcast = f"""NOTIFY * HTTP/1.1
```
Comment (sanderjo) [inline sabnzbd/utils/ssdp.py]: @Safihre I could move the creation of the SSDP and the XML to the __init__(). That way they are only created once, and the variables used are fixed after __init__ has done its work ...

Comment (sanderjo) [reply sabnzbd/utils/ssdp.py]: ... just changed the code like that.

File: sabnzbd/utils/ssdp.py [hunk 5]
```diff
@@ -34,9 +41,45 @@ def __init__(self, host, server_name, url, description):
...
+        self.__uuid = uuid.uuid3(uuid.NAMESPACE_DNS, self.__myhostname + self.__host)
+
+        # descriptionxmlURL = self.__url + "/description.xml"
+
+        # Create the SSDP broadcast message
+        self.__mySSDPbroadcast = f"""NOTIFY * HTTP/1.1
+HOST: 239.255.255.250:1900
+CACHE-CONTROL: max-age=60
+LOCATION: {self.__url}/description.xml
+SERVER: SABnzbd
```
Comment (Safihre) [inline sabnzbd/utils/ssdp.py:54]: Use `self.__server_name`

File: sabnzbd/utils/ssdp.py [hunk 6]
```diff
@@ -34,9 +41,45 @@ def __init__(self, host, server_name, url, description):
         self.__url = url
         self.__description = description
         self.__myhostname = socket.gethostname()
-        # uuid stays the same as long as hostname and ip address stay the same:
-        self.__uuidXML = uuid.uuid3(uuid.NAMESPACE_DNS, self.__myhostname + self.__host)
-        self.__uuidSSDP = uuid.uuid3(uuid.NAMESPACE_DNS, self.__host + self.__myhostname)
+        # a steady uuid: stays the same as long as hostname and ip address stay the same:
+        self.__uuid = uuid.uuid3(uuid.NAMESPACE_DNS, self.__myhostname + self.__host)
+
+        # descriptionxmlURL = self.__url + "/description.xml"
```
Comment (Safihre) [inline sabnzbd/utils/ssdp.py:47]: Remove.

File: sabnzbd/utils/ssdp.py [hunk 7]
```diff
@@ -34,9 +41,45 @@ def __init__(self, host, server_name, url, description):
...
+LOCATION: {self.__url}/description.xml
+SERVER: SABnzbd
+NT: upnp:rootdevice
+USN: uuid:{self.__uuid}::upnp:rootdevice
+NTS: ssdp:alive
+OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
+
+"""
+        self.__mySSDPbroadcast = self.__mySSDPbroadcast.replace("\n", "\r\n")
+        self.__mySSDPbroadcast = bytes(self.__mySSDPbroadcast, "utf-8")  # convert string to bytes
```
Comment (Safihre) [inline sabnzbd/utils/ssdp.py:62]: The Python way to do this is ```self.__mySSDPbroadcast.encode("utf-8")```.

File: sabnzbd/utils/ssdp.py [hunk 8]
```diff
@@ -34,9 +41,45 @@ def __init__(self, host, server_name, url, description):
...
+        self.__myxml = f"""<?xml version="1.0" encoding="UTF-8" ?>
+<root xmlns="urn:schemas-upnp-org:device-1-0">
+<specVersion>
+<major>1</major>
+<minor>0</minor>
+</specVersion>
+<URLBase>{self.__url}</URLBase>
+<device>
+<deviceType>urn:schemas-upnp-org:device:Basic:1</deviceType>
+<friendlyName>SABnzbd ({self.__myhostname})</friendlyName>
```
Comment (Safihre) [inline sabnzbd/utils/ssdp.py:74]: Use `self.__server_name`

File: sabnzbd/utils/ssdp.py (hunk 5 above)
Comment (sanderjo) [reply sabnzbd/utils/ssdp.py:54]: It must be a URL. server_name is just "SABnzbd". So, IMHO, won't work

image

File: sabnzbd/utils/ssdp.py [hunk 9]
```diff
@@ -34,9 +41,45 @@ def __init__(self, host, server_name, url, description):
...
+<manufacturer>SABnzbd Team</manufacturer>
+<manufacturerURL>http://www.sabnzbd.org</manufacturerURL>
```
Comment (Safihre) [inline sabnzbd/utils/ssdp.py:76]: We need to get rid of all the SABnzbd stuff. I would suggest adding 3 extra fields to `__init__`:
```
__init__(self, host, server_name, url, description, manufacturer, manufacturer_url, model)
```
And make the respective `self.__x` variables.

**Review by Safihre — CHANGES_REQUESTED** [review]

**Comment by Safihre** [comment]
While testing and putting my mac to sleep I did get this crash:
```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/Users/xxx/Documents/GitHub/sabnzbd/sabnzbd/utils/ssdp.py", line 108, in run
    sock.sendto(self.__mySSDPbroadcast, (MCAST_GRP, MCAST_PORT))
OSError: [Errno 50] Network is down
```
Maybe we should add a try/except

**Comment by sanderjo** [comment]
> Maybe we should add a try/except

Confirmed & Done!

**Pull request closed without merging**